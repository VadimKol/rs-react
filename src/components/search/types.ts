export interface SearchProps {
  searchField: React.RefObject<HTMLInputElement>;
  character: { name: string };
  setCharacter: (character: { name: string }) => void;
  loader: boolean;
}
