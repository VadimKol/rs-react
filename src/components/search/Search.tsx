import { type ReactNode } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';

import { useLocalStorage } from '@/hooks/useLocalStorage';

import styles from './styles.module.scss';
import type { SearchProps } from './types';

export function Search({ searchField, character, setCharacter, setPage, loader }: SearchProps): ReactNode {
  const [, setPageQuery] = useSearchParams();
  const [, setLs] = useLocalStorage('R&M_search');
  const navigate = useNavigate();

  return (
    <form
      className={styles.search_form}
      onSubmit={(e) => {
        e.preventDefault();
        if (typeof searchField.current?.value === 'string' && !loader) {
          const searchValue = searchField.current?.value.trim();
          setCharacter({ name: searchValue });
          setLs(searchValue);
          setPage(1);
          setPageQuery({ page: '1' });
          navigate('/', { replace: true });
        }
      }}
    >
      <input
        id="search"
        className={styles.search}
        type="text"
        placeholder="Search..."
        ref={searchField}
        defaultValue={character.name}
      />
      <button type="submit" className={styles.search_button} aria-label="Search-button" />
    </form>
  );
}
