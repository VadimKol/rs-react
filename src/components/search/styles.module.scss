.search_form {
  display: flex;
  justify-content: center;

  & .search_box {
    position: relative;
  }

  & .search {
    box-sizing: border-box;
    width: 400px;
    height: 45px;
    padding: 0 40px 0 5px;
    font-family: var(--primary-font);
    font-size: 30px;
    color: var(--text-color);
    background: var(--background-color) padding-box;
    border: 2.5px solid transparent;
    border-radius: 6px;

    &:focus {
      border: 2.5px solid var(--primary-color);
    }

    &:focus-visible {
      outline: none;
    }
  }

  & .search_button {
    position: absolute;
    top: 7px;
    right: 7px;
    width: 30px;
    height: 30px;
    padding: 0;
    background-color: var(--text-color);
    border: none;
    transition: background-color 0.5s;
    mask-image: url("/src/assets/images/search.svg");
    mask-size: cover;

    &:focus-visible {
      border-radius: 6px;
      outline: 2px solid var(--primary-color);
    }

    @media (hover: hover) {
      &:hover {
        background-color: var(--primary-color);
        cursor: pointer;
      }
    }
  }
}
