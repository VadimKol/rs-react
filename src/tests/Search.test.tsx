import { createRemixStub } from '@remix-run/testing';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { type ReactNode } from 'react';

import { Search } from '@/components/search/Search';

const navigate = vi.fn((arg: string): string => arg);

describe('Search Component', () => {
  const user = userEvent.setup();

  it('renders correctly', () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: (): ReactNode => <Search loader={false} />,
      },
    ]);

    const { container } = render(<RemixStub />);

    expect(container).toMatchSnapshot();
  });

  it('checks search value in URL', async () => {
    vi.mock('@remix-run/react', () => ({
      useSearchParams: vi.fn(() => [new URLSearchParams('')]),
      useNavigate: vi.fn(() => navigate),
    }));

    render(<Search loader={false} />);

    const buttonSearch = screen.getByRole('button', { name: 'Search-button' });
    const inputSearch = screen.getByPlaceholderText('Search...');

    await user.type(inputSearch, 'Rick');
    await user.click(buttonSearch);

    expect(navigate).toHaveBeenCalledWith(`/?page=1&search=Rick`);
  });
});
