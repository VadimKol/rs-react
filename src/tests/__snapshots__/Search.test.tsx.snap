// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Component > renders correctly 1`] = `
<div>
  <form
    class="_search_form_5ff5ab"
  >
    <input
      class="_search_5ff5ab"
      id="search"
      placeholder="Search..."
      type="text"
      value=""
    />
    <button
      aria-label="Search-button"
      class="_search_button_5ff5ab"
      type="submit"
    />
  </form>
</div>
`;
